### PR TITLE
#352 Wild card, supplemental messages support

### PR DESCRIPTION
**Pull request by cragun47**
Signed-off-by: Brian Cragun <cragun@us.ibm.com>

Initial drop of Wild card, message discovery, supplemental messages support.
Documentation in future drop.

### CODE REVIEW

File: src/test/resources/additional_resources/hl7/message/ADT_A09.yml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#
```
Comment (LisaWellman) [inline src/test/resources/additional_resources/hl7/message/ADT_A09.yml:1]: Did you man to deliver this?

**Review by LisaWellman — COMMENTED** [review]
Have not looked at HL7ResourceReaderTest but wanted to get some of my comments out there.

File: src/test/java/io/github/linuxforhealth/hl7/resource/HL7ResourceReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    // AND that they are found when supported.hl7.messages is omitted and defaults to *
+    @Test
+    public void testGetMessageTemplatesViaAdditionalLocationWithDefaultSupportedList() throws IOException {
+      try {
+        // Set up the config file
+        File configFile = new File(folder, "config.properties");
+        Properties prop = new Properties();
+        prop.put("base.path.resource", "src/main/resources");
+        prop.put("default.zoneid", "+08:00");
+        prop.put("additional.resources.location", "src/test/resources/additional_resources");
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/hl7/resource/HL7ResourceReaderTest.java:110]: Add a comment here in the code that supported.hl7.messages is omitted on purpose, so no one adds it (and doesn't read the method comment)

File: README.md [hunk 3]
```diff
@@ -102,10 +102,11 @@ The converter configuration file, config.properties, supports the following sett
...
+| supported.hl7.messages  | Comma delimited list of hl7 message/event types. An asterisk `*` may be used to indicate all messages found in sub-directory `/hl7/messages` under the `base.path.resource` and sub-directory `/hl7/messages` under `additional.resources.location` are supported. If not specified, defaults to `*`.                                                                                                                             | ADT_A01, ORU_R01, PPR_PC1       |
 | default.zoneid          | ISO 8601 timezone offset (optional). The zoneid is applied to translations when the target FHIR resource field requires a timezone, but the source HL7 field does not include it. | +08:00                          |
 | additional.conceptmap   | Path to additional concept map configuration. Concept maps are used for mapping one code system to another.                                                                       | /opt/converter/concept-map.yaml |
+| additional.resources.location  | Path to additional resources. These supplement those `base.path.resource`.                                                                         | /opt/converter/concept-map.yaml |
```
[the remark this answers is not in the document]
Comment (cragun47) [reply README.md:109]: Done

File: src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java [hunk 4]
```diff
@@ -107,12 +122,47 @@ public String getResource(String resourcePath) {
     return messagetemplates;
   }
 
+  private boolean hasWildcard(List<String> supportedMessageTemplates) {
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java:125]: Noted.

File: src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java [hunk 5]
```diff
@@ -107,12 +122,47 @@ public String getResource(String resourcePath) {
...
+        foundTemplates.add(listOfFiles[i].getName());
+      }
+    }
+
+    return foundTemplates;
+  }
+
+  private HL7MessageModel getMessageModel(String templateName) {
+    // Allow for names that already have .yml extension
+    String yamlizedTemplateName = templateName.contains(".yml") ? templateName : templateName + ".yml";
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java:164]: Easy change.  Done.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java [hunk 6]
```diff
@@ -67,11 +67,43 @@ public void test_that_additional_conceptmap_values_are_loaded() throws IOExcepti
         assertThat(url).isEqualTo("http://hl7.org/fhir/StructureDefinition/patient-mothersMaidenName");
     }
 
+    @Test
+    public void test_that_additional_resources_location_is_found() throws IOException {
+    	File configFile = new File(folder, "config.properties");
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:72]: This file comes from 
```jave
	@TempDir
    static File folder;
```

The scope is only for the life of the test, so it automatically cleans up.
https://junit.org/junit5/docs/5.4.2/api/org/junit/jupiter/api/io/TempDir.html

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java [hunk 7]
```diff
@@ -67,11 +67,43 @@ public void test_that_additional_conceptmap_values_are_loaded() throws IOExcepti
...
+        writeProperties(configFile);
+        System.setProperty(CONF_PROP_HOME, configFile.getParent());
+        ConverterConfiguration.reset();
+        ConverterConfiguration theConvConfig = ConverterConfiguration.getInstance();
+        assertThat(theConvConfig.getSupportedMessageTemplates()).hasSize(4);
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:77]: The four files are from the config file that we build and introduce.  The default config file is the one used for converting, and it has a dozen templates.  Thus this test does what it's supposed to  and loads just the four.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java [hunk 8]
```diff
@@ -67,11 +67,43 @@ public void test_that_additional_conceptmap_values_are_loaded() throws IOExcepti
...
         prop.put("default.zoneid", "+08:00");
         prop.put("additional.conceptmap.file", "src/test/resources/additional_conceptmap.yml");
+        prop.put("additional.resources.location", "src/test/resources/additional_resources");
+        prop.store(new FileOutputStream(configFile), null);
+    }
+
+
+    @Test
+    public void test_that_supportedhl7messages_defaults_to_asterisk() throws IOException {
+    	File configFile = new File(folder, "config.properties");
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:93]: Same answer as above:
This file comes from
```java
	@TempDir
    static File folder;
```
The scope is only for the life of the test, so it automatically cleans up.
https://junit.org/junit5/docs/5.4.2/api/org/junit/jupiter/api/io/TempDir.html

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply README.md:109]: Change made.

File: src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java [hunk 9]
```diff
@@ -71,20 +71,24 @@ private static String loadClassPathResource(String resourceConfigurationPath) th
...
+    Path resourceFolderFilePath = Paths.get(converterConfig.getResourceFolder(), resourcePath);
+    Path alternateResourceFolderFilePath = Paths.get(converterConfig.getAdditionalResourcesLocation(), resourcePath);
     String resource;
 
     try {
-      if (filePath != null && filePath.toFile().exists()) {
-        resource = loadFileResource(filePath.toFile());
+      if (resourceFolderFilePath != null && resourceFolderFilePath.toFile().exists()) {
+        resource = loadFileResource(resourceFolderFilePath.toFile());
+      } else if (alternateResourceFolderFilePath != null && alternateResourceFolderFilePath.toFile().exists()) {
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java:90]: OK.  Thanks.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:93]: Done

File: src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java [hunk 10]
```diff
@@ -107,12 +122,47 @@ public String getResource(String resourcePath) {
...
+    }
 
-    String templateFileContent =
-        getResourceInHl7Folder(Constants.MESSAGE_BASE_PATH + templateName + ".yml");
+    // Add the extended templates
+    // Current code assumes the extended templates are exclusive of the base templates.
+    // In future, if there is to be a priority, or one can override the other, changes needed here to merge extended templates.
+    folder = new File(
+        converterConfig.getAdditionalResourcesLocation() + '/' + Constants.HL7_BASE_PATH + Constants.MESSAGE_BASE_PATH);
+    listOfFiles = folder.listFiles();
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/main/java/io/github/linuxforhealth/hl7/resource/ResourceReader.java:152]: I added comments for clarification.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:77]: Yes.  As we discussed this is a real message, but one that we won't use.  It is in a testing directory.  Added comments for clarity.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:72]: This is just testing the config, not the way the config is used.
Added comments for clarity and added additional useful tests.

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java [hunk 11]
```diff
@@ -67,11 +67,43 @@ public void test_that_additional_conceptmap_values_are_loaded() throws IOExcepti
...
+        writePropertiesDefaultMessages(configFile);
+        System.setProperty(CONF_PROP_HOME, configFile.getParent());
+        ConverterConfiguration.reset();
+        ConverterConfiguration theConvConfig = ConverterConfiguration.getInstance();
```
[the remark this answers is not in the document]
Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:97]: Done.

**Review by LisaWellman — APPROVED** [review]

File: src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java [hunk 12]
```diff
@@ -68,42 +68,47 @@ public void test_that_additional_conceptmap_values_are_loaded() throws IOExcepti
...
+    public void testConfigurationDefaultsAreUsed() throws IOException {
     	File configFile = new File(folder, "config.properties");
         writePropertiesDefaultMessages(configFile);
         System.setProperty(CONF_PROP_HOME, configFile.getParent());
         ConverterConfiguration.reset();
         ConverterConfiguration theConvConfig = ConverterConfiguration.getInstance();
-        assertThat(theConvConfig.getSupportedMessageTemplates()).hasSize(1);
-        assertThat(theConvConfig.getSupportedMessageTemplates().get(0)).contains("*");
+        assertThat(theConvConfig.getResourceFolder()).isEmpty(); // No resource folder specified
+        assertThat(theConvConfig.getSupportedMessageTemplates()).hasSize(1); // Because there was an *, there is only only message template configured
```
Comment (LisaWellman) [inline src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java:103]: nit - "only only" should be "only one"

Comment (cragun47) [reply src/test/java/io/github/linuxforhealth/core/config/ConverterConfigurationTest.java]: Easy fix.  Done.

**Pull request merged**